Always convert to Int64 in difference(), intersect(), union()

hagenw opened the pull request:
Closes #272 

This adjust the return type for `audformat.utils.difference()`, `audformat.utils.intersect()`, `audformat.utils.union()`
to always return `Int64` for integers.

![image](https://user-images.githubusercontent.com/173624/182647183-07bb1d48-c353-4dab-98d8-3ab3c664494e.png)

![image](https://user-images.githubusercontent.com/173624/182647240-825c65e7-0c1f-479b-b380-31463b7ff902.png)

![image](https://user-images.githubusercontent.com/173624/182647300-877225b3-58c8-4a83-b168-eff5e4e96930.png)

hagenw commented:
> Closes #272

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -395,7 +393,15 @@ def difference(
-    index = _alike_index(objs[0], index)
+    # select object with highest dtype for integers
+    alike_obj = objs[0]
+    if 'int64' in _dtypes(alike_obj):
```
What if we have two integer levels but the object that is now selected has only `Int64` in one of them, whereas another one has it in the other place?

frankenjoe replied:
I wonder if we should simply always convert integer levels to `Int64` in the result of `union()`, `intersect()` and `difference()`?

hagenw replied:
> What if we have two integer levels but the object that is now selected has only Int64 in one of them, whereas another one has it in the other place?

You mean something like this:
```python
index = pd.MultiIndex.from_arrays([[0], [1]], names=['idx1', 'idx2'])
index1 = audformat.utils.set_index_dtypes(index, {'idx1': 'Int64', 'idx2': 'int64'})
index2 = audformat.utils.set_index_dtypes(index, {'idx1': 'int64', 'idx2': 'Int64'})
audformat.utils.difference([index1, index2]).dtypes
```
There we expect that it returns
```
idx1    Int64
idx2    Int64
dtype: object
```
but what we get is
```python
idx1    int64
idx2    Int64
dtype: object
```

frankenjoe replied:
yes

frankenjoe replied:
We can probably fix this, but code will become quite ugly. So always converting to `Int64` makes sense I think. Also because we anyway convert to `Int64` with `Table` and `MiscTable`.

hagenw replied:
OK, I changed to always converting to `Int64`.

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -1730,6 +1728,24 @@ def _is_same_dtype(d1, d2) -> bool:
     return d1.name == d2.name
 
 
+def _maybe_convert_int_dtype(
```
We have more or less the same code in `MiscTable.__init__()` so maybe we can call this function there, too?

hagenw replied:
Not so easy as we need access to `levels` and `dtypes` after the code block.

hagenw replied:
OK, I reuse the helper function now.

frankenjoe commented:
I think in `intersect()` we still have one case where we do not convert:

```python
    # return if the shortest obj has no entries
    if len(objs_sorted[0]) == 0:
        return _alike_index(objs[0])
```

frankenjoe commented:
And then we also have this part:

```python
    # Ensure we have order of first object
    index = objs[0].intersection(index)
    if isinstance(index, pd.MultiIndex):
        index = set_index_dtypes(index, objs[0].dtypes.to_dict())
```

Couldn't this revert back to `int64`?

frankenjoe commented:
To not miss a case by accident, maybe it would be simpler to always start the functions with:

```python
objs = [_maybe_convert_int_dtype(obj) for obj in objs]
```

?

hagenw commented:
> To not miss a case by accident, maybe it would be simpler to always start the functions with:
> 
> ```python
> objs = [_maybe_convert_int_dtype(obj) for obj in objs]
> ```
> 
> ?

I tried this, this will lead to several other failures ;)

frankenjoe commented:
> > 
> 
> I tried this, this will lead to several other failures ;)

Do we know why?

hagenw commented:
> ```python
> objs = [_maybe_convert_int_dtype(obj) for obj in objs]
> ```

Right, I added a failing test in https://github.com/audeering/audformat/pull/273/commits/020c30fc59737a58928bd451a9e01bcc800725a4 and fixed it in https://github.com/audeering/audformat/pull/273/commits/0764c29efdeacad01f7f715e09938615859051aa

hagenw commented:
> And then we also have this part:
> 
> ```python
>     # Ensure we have order of first object
>     index = objs[0].intersection(index)
>     if isinstance(index, pd.MultiIndex):
>         index = set_index_dtypes(index, objs[0].dtypes.to_dict())
> ```
> 
> Couldn't this revert back to `int64`?

Yes, tests were not failing, because some expected the wrong output. I fixed the test in https://github.com/audeering/audformat/pull/273/commits/95f2ce4fde3e73081c4a8512eb4a78f54a18bedf and changed the code in https://github.com/audeering/audformat/pull/273/commits/31f1a3acfd5e3f927f8dc68515b422ff2a2229fb

hagenw commented:
> To not miss a case by accident, maybe it would be simpler to always start the functions with:
> 
> ```python
> objs = [_maybe_convert_int_dtype(obj) for obj in objs]
> ```
> 
> ?

You were right, this works. I changed the code accordingly now.

What was failing for me before was only changing the first object.
